javascript/pdpModule.js

```diff
@@ -0,0 +1,170 @@
+        <span>
+            <p>Subtotal</p>
+            <p> &#8377 ${cartItems.subTotal}</p>
+        </span>
+        <p class="charges-description">Extra charges may apply</p>
+        <button class="checkout-btn">Checkout &#8594 </button>
+        </div>`;
+    }
+
+    const categories = fetchCategories();
+
+    const menuItems = fetchMenuItems();
+
+    const cartItems = fetchCartItems();
+
+    function addItemByCategoryInMap(map, key, menuItem) {
+        if (!map.has(key)) {
+            map.set(key, []);
+        }
+        map.get(key).push(menuItem);
+    }
+
+    function addMenuItemsInMap(map, menuItem) {
+        menuItem.categories.forEach(key => addItemByCategoryInMap(map, key, menuItem));
+    }
+
+    function createItemByCategoryMap(menuItems) {
+        const map = new Map();
+        menuItems.forEach(menuItem => addMenuItemsInMap(map, menuItem));
+        return map;
+    }
+
+    const itemsByCategoryMap = createItemByCategoryMap(menuItems);
+
+    const main = document.querySelector("main");
+
+    const categoriesContainer = document.querySelector(".categories");
+
+    const categoriesList = document.createElement("ul");
+
+    function addListItemInCategoryList(listElement) {
+        const categoriesListElement = document.createElement("li");
+        categoriesListElement.id = listElement.id;
+        categoriesListElement.innerText = listElement.displayName;
+        categoriesList.append(categoriesListElement);
+    }
+    categories.forEach(listElement => addListItemInCategoryList(listElement));
+    categoriesContainer.append(categoriesList);
+
+    const menuItemsContainer = document.querySelector(".menu-items");
+
+    function addItemToCartegory(itemsByCategoryMap, key, index) {
+        const itemContainer = document.createElement("div");
+        itemContainer.className = "item-description";
+        itemContainer.innerHTML = createItemTemplate(itemsByCategoryMap, key, index);
+
+        menuItemsContainer.append(itemContainer);
+    }
+
+    function addMenuItemsToCategory(itemsByCategoryMap, key) {
+        const categoryInfoContainer = document.createElement("div");
+        categoryInfoContainer.className = "category-info";
+        categoryInfoContainer.innerHTML = addCategoryInfoTemplate(itemsByCategoryMap, key);
+        menuItemsContainer.append(categoryInfoContainer);
+        for (let i = 0; i < itemsByCategoryMap.get(key).length; i++) {
+            addItemToCartegory(itemsByCategoryMap, key, i);
+        }
+    }
+
+    function addCategoriesToMenuList(itemsByCategoryMap) {
+        for (const [key, value] of itemsByCategoryMap.entries()) {
+            addMenuItemsToCategory(itemsByCategoryMap, key);
+        }
+    };
+
+    addCategoriesToMenuList(itemsByCategoryMap);
+
+    const cartContainer = document.querySelector(".cart");
+    cartContainer.innerHTML = createCartDescriptionTemplate(cartItems);
+    main.append(categoriesContainer, menuItemsContainer, cartContainer);
```

Comment:
Always expose a public API to the consumer based on which the consumer can decide on when to initiate the module.

Right now, this will just get called immediately on load. If later, you had to wait for some user action or render this module conditionally - you will not be able to do so.

An example scenario: Let's say you need to check whether the user is Sector 29 or Sector 55 and based on the location want to give out different offers. To do that you would have to pass the required parameters while initialising the module.

Expose the rendering logic as `init` method from this module and then call the init method to initiate rendering.